ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/BulkSync.hs

```diff
@@ -0,0 +1,344 @@
+
+-- | Given a list of candidate fragments and their associated peers, choose what
+-- to sync from who in the bulk sync mode.
+fetchDecisionsBulkSync ::
+  ( HasHeader header,
+    HeaderHash header ~ HeaderHash block,
+    Eq peer
+  ) =>
+  FetchDecisionPolicy header ->
+  -- | The current chain, anchored at the immutable tip.
+  AnchoredFragment header ->
+  (Point block -> Bool) ->
+  MaxSlotNo ->
+  PeersOrder peer ->
+  -- | Association list of the candidate fragments and their associated peers.
+  -- The candidate fragments are anchored in the current chain (not necessarily
+  -- at the tip; and not necessarily forking off immediately).
+  [(AnchoredFragment header, PeerInfo header peer extra)] ->
+  -- | Association list of the requests and their associated peers. There is at
+  -- most one accepted request; everything else is declined. Morally, this is a
+  -- map from peers to @'FetchDecision' ('FetchRequest' header)@ with at most
+  -- one @'FetchRequest' header@.
+  ( Maybe (FetchRequest header, PeerInfo header peer extra),
+    [(FetchDecline, PeerInfo header peer extra)]
+  )
+fetchDecisionsBulkSync
+  fetchDecisionPolicy
+  currentChain
+  fetchedBlocks
+  fetchedMaxSlotNo
+  peersOrder
+  candidatesAndPeers = combineWithDeclined $ do
+    -- Step 1: Select the candidate to sync from. This already eliminates peers
+    -- that have an implausible candidate. It returns the remaining candidates
+    -- (with their corresponding peer) as suffixes of the immutable tip.
+    ( theCandidate :: ChainSuffix header,
+      candidatesAndPeers' :: [(ChainSuffix header, PeerInfo header peer extra)]
+      ) <-
+      MaybeT $
+        selectTheCandidate
+          fetchDecisionPolicy
+          currentChain
+          candidatesAndPeers
+
+    -- Step 2: Select the peer to sync from. This eliminates peers that cannot
+    -- serve a reasonable batch of the candidate, then chooses the peer to sync
+    -- from, then again declines the others.
+    ( thePeerCandidate :: ChainSuffix header,
+      thePeer :: PeerInfo header peer extra
+      ) <-
+      MaybeT $
+        selectThePeer
+          fetchDecisionPolicy
+          fetchedBlocks
+          fetchedMaxSlotNo
+          peersOrder
+          theCandidate
+          candidatesAndPeers'
+
+    -- Step 3: Fetch the candidate from the selected peer, potentially declining
+    -- it (eg. if the peer is already too busy).
+    MaybeT $
+      fetchTheCandidate
+        fetchDecisionPolicy
+        fetchedBlocks
+        fetchedMaxSlotNo
+        theCandidate
+        thePeer
+        thePeerCandidate
+    where
+      combineWithDeclined ::
+        MaybeT (WithDeclined peer) (a, peer) ->
+        ( Maybe (a, peer),
+          [(FetchDecline, peer)]
+        )
+      combineWithDeclined = second listConcatToList . runWithDeclined . runMaybeT
+
+-- FIXME: The 'FetchDeclineConcurrencyLimit' should only be used for
+-- 'FetchModeDeadline', and 'FetchModeBulkSync' should have its own reasons.
```

Comment:
It still needs to be fixed, but I am not sure if I actually think that it would need fixing, I am just wondering about clarity. The peers are indeed declined because we limit ourselves to a concurrency of 1, so it is a pretty degenerate case of this constructor and I wouldn't want it to mixed things up. It is probably reasonable to keep things as-is.

Reply:
Removed it. I think `FetchDeclineConcurrencyLimit` is fine. If someone finds it annoying later on in a debugging session, then they can of course refine it.